src/crypto/sha256_sse41.cpp

```diff
@@ -10,6 +14,221 @@
+;   documentation and/or other materials provided with the
+;   distribution.
+;
+; * Neither the name of the Intel Corporation nor the names of its
+;   contributors may be used to endorse or promote products derived from
+;   this software without specific prior written permission.
+;
+;
+; THIS SOFTWARE IS PROVIDED BY INTEL CORPORATION "AS IS" AND ANY
+; EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+; IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+; PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL INTEL CORPORATION OR
+; CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+; EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+; PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+; PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+; LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+;
+; This code is described in an Intel White-Paper:
+; "Fast SHA-256 Implementations on Intel Architecture Processors"
+;
+; To find it, surf to http://www.intel.com/p/en_US/embedded
+; and search for that title.
+; The paper is expected to be released roughly at the end of April, 2012
+;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+; This code schedules 1 blocks at a time, with 4 lanes per block
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+*/
+
+namespace {
+
+uint32_t inline __attribute__((always_inline)) Ror(uint32_t x, int val) { return ((x >> val) | (x << (32 - val))); }
+
+void inline __attribute__((always_inline)) Round(uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d, uint32_t& e, uint32_t& f, uint32_t& g, uint32_t& h, uint32_t w)
+{
+    uint32_t y0, y1, y2;
+    y0 = Ror(e, 25 - 11) ^ e;
+    y1 = Ror(a, 22 - 13) ^ a;
+    y2 = (f ^ g) & e;
+    y0 = Ror(y0, 11 - 6) ^ e;
+    y1 = Ror(y1, 13 - 2) ^ a;
+    y0 = Ror(y0, 6);
+    y2 = (y2 ^ g) + y0 + w;
+    y1 = Ror(y1, 2);
+    h += y2;
+    y0 = (a | c) & b;
+    d += h;
+    h += y1;
+    y0 |= (a & c);
+    h += y0;
+}
+
+void inline __attribute__((always_inline)) QuadRoundSched(uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d, uint32_t& e, uint32_t& f, uint32_t& g, uint32_t& h, __m128i& X0, __m128i X1, __m128i X2, __m128i X3, __m128i W)
+{
+    alignas(__m128i) uint32_t Ws[4];
+    __m128i XTMP0, XTMP1, XTMP2, XTMP3, XTMP4;
+    const __m128i SHUF_00BA = _mm_set_epi64x(0xFFFFFFFFFFFFFFFFULL, 0x0b0a090803020100ULL);
+    const __m128i SHUF_DC00 = _mm_set_epi64x(0x0b0a090803020100ULL, 0xFFFFFFFFFFFFFFFFULL);
+
+    W = _mm_add_epi32(W, X0);
+    _mm_store_si128((__m128i*)Ws, W);
+
+    Round(a, b, c, d, e, f, g, h, Ws[0]);
+    XTMP0 = _mm_alignr_epi8(X3, X2, 4);
+    XTMP0 = _mm_add_epi32(XTMP0, X0);
+    XTMP3 = XTMP2 = XTMP1 = _mm_alignr_epi8(X1, X0, 4);
```

Comment:
Is there some voodoo here, why not just use XTMP3 below? Does this avoid a pipeline stall or something?

Reply:
No idea. It's just a translation of the existing assembly code.